### PR TITLE
Profile.print: color Base/Core & packages. Make paths clickable

### PR DESCRIPTION
**Pull request by IanButterworth**
Updated
## This PR
Screenshot 2024-09-02 at 1 47 23 PM

## master
Screenshot 2024-09-02 at 1 49 42 PM

Todo:
- [ ] ~Maybe drop the `@` prefix when coloring it, given it's obviously special when colored~ If someone copy-pasted the profile into an issue this would make it confusing.
- [ ] Figure out why `Profile.print(format=:flat)` is truncating before the terminal width is used up
- [x] Make filepaths terminal links (even if they're truncated)

### CODE REVIEW

**Comment by KristofferC** [comment]
Orthogonal but the `@Module` syntax prevents you from Ctrl-clicking them in e.g. VSCode to take you to the line but if we made them terminal links that functionality would probably be restored.

**Comment by tecosaur** [comment]
> if we made them terminal links

FYI I've already got code for that in #51816.

**Comment by tecosaur** [comment]
IMO [the commit that adds `uripath`](https://github.com/JuliaLang/julia/pull/51816/commits/780300cb3c6c0ea4fb8b2af309e54540a06e9ca8) could be landed separately, I'm sort of waiting for the StyledStrings/Base printing type piracy to get a bit closer to resolved before pushing that PR again.

Happy to make that PR if that would be helpful, or we can just wait.

**Comment by IanButterworth** [comment]
Sounds good. Doesn't need to hold this up though. 

@tecosaur would you mind reviewing for whether there's an easy way to implement this better with annotated strings stuff? If it requires an overhaul then maybe that could be a following PR.

File: stdlib/Profile/src/Profile.jl [hunk 1]
```diff
@@ -924,16 +937,20 @@ function tree_format(frames::Vector{<:StackFrameTree}, level::Int, cols::Int, ma
...
+                pkgcolor = get!(() -> popfirst!(Base.STACKTRACE_MODULECOLORS), PACKAGE_FIXEDCOLORS, pkgname)
+                remaining_path = rtruncto(filename, widthfile - length(pkgname) - 1)
                 strs[i] = string(stroverhead, "╎", base, strcount, " ",
-                    rtruncto(filename, widthfile),
+                    sprint((io,x) -> printstyled(io, x; color=pkgcolor), pkgname; context=io),
+                    !isempty(pkgname) && !startswith(remaining_path, "/") ? "/" : "",
+                    remaining_path,
                     ":",
                     li.line == -1 ? "?" : string(li.line),
                     "; ",
```
Comment (tecosaur) [inline stdlib/Profile/src/Profile.jl:956]: Ok, GitHub doesn't seem to let me comment on lines more than three lines away from the diff changes, but this is actually pursuant to line 945 (962 with your patch):

```julia
strs[i] = ltruncto(strs[i], cols)
```

There are two problems here, one existing, one introduced.

Firstly, the existing problem: `ltruncto` / `rtruncto` mix up indices and `textwidth` and that means that `textwidth(ltruncto(str, n)) <= n` doesn't always hold. For example `textwidth(ltruncto("🍕" ^ 15))` is `19`.

Now the new problem: by introducing `sprint` + `printstyled` above, `strs[i]` now contains ANSI escape codes that are counted by `length`/`textwidth` but don't actually increase the width of the string. This is going to cause lines to be systematically over-shortened.

```julia-repl
julia> textwidth(sprint(io -> printstyled(io, "hey"), context=stdout))
6
```

**Comment by IanButterworth** [comment]
File paths are now clickable in any terminal that supports it, even if they are truncated. See example up top.

I opted to use IDE-specific formats over a generic URI because the generic URI format doesn't seem to support linenums.

**Comment by gdalle** [comment]
Looks much more usable! I wonder if coloring the whole file path instead of just the module would make things even clearer?

**Comment by IanButterworth** [comment]
@gdalle interesting idea..

Screenshot 2024-09-02 at 7 45 27 AM

The underline styling of links in iTerm2 makes it a little more busy than it needs to be, but I think its clearer

**Comment by IanButterworth** [comment]
I think the coloring makes the trailing `;` superfluous now.

**Comment by lgoettgens** [comment]
With some of the colors, I find it incredibly hard to read what's really written there. For the modules, this is fine, but for the paths this could be avoided.

**Comment by gbaraldi** [comment]
One idea (which might be bad) would be to only colour the stacktraces that are heavier? Something like at least 5% of the samples

**Comment by IanButterworth** [comment]
Maybe bolding lines that have numbers on the far left would help.

**Comment by IanButterworth** [comment]
Yeah I think that helps

Screenshot 2024-09-02 at 1 29 03 PM

**Comment by IanButterworth** [comment]
I think the `;` after the path is unnecessary and looks cleaner with just a double space.

Screenshot 2024-09-02 at 1 47 23 PM

File: stdlib/Profile/src/Allocs.jl [hunk 2]
```diff
@@ -321,7 +321,7 @@ end
 function flat(io::IO, data::Vector{Alloc}, cols::Int, fmt::ProfileFormat)
     fmt.combine || error(ArgumentError("combine=false"))
     lilist, n, m, totalbytes = parse_flat(fmt.combine ? StackFrame : UInt64, data, fmt.C)
-    filenamemap = Dict{Symbol,String}()
+    filenamemap = FileNameMap()
```
Comment (IanButterworth) [inline stdlib/Profile/src/Allocs.jl:324]:
```suggestion
    filenamemap = Profile.FileNameMap()
```

**Comment by IanButterworth** [comment]
I'm marking this for merge as I think the formatting changes have support, and the clickable links part will need some testing in the wild

File: stdlib/Profile/src/Profile.jl [hunk 3]
```diff
@@ -788,9 +794,34 @@ function flat(io::IO, data::Vector{UInt64}, lidict::Union{LineInfoDict, LineInfo
     return false
 end
 
+# make a terminal-clickable link to the file and linenum.
+# Similar to `define_default_editors` in `Base.Filesystem` but for creating URIs not commands
+function editor_link(path::String, linenum::Int)
```
Comment (tecosaur) [inline stdlib/Profile/src/Profile.jl:799]: At a glance it looks like `path` is directly interpolated? This looks a little dodgy to me, what about a path like `/tmp/?file=redherring&line=42/file.jl`? Will that become say `idea://open?file=/tmp/?file=redherring&line=42/file.jl&line=1`

Comment (IanButterworth) [reply stdlib/Profile/src/Profile.jl:799]: That path isn't a file path? and especially not something that the profiler could return, as far as I understand...?
Though escaping the path generally seems like a good idea.

Also it'd be great to have https://github.com/JuliaLang/julia/pull/55454 finished to use in this function for the generic case.

Comment (IanButterworth) [reply stdlib/Profile/src/Profile.jl:799]: Ah I see what you're saying.. a dir named `?file=redherring&line=42`. Ok

Comment (IanButterworth) [reply stdlib/Profile/src/Profile.jl:799]: https://github.com/JuliaLang/julia/pull/55857

Comment (tecosaur) [reply stdlib/Profile/src/Profile.jl:799]: > Also it'd be great to have https://github.com/JuliaLang/julia/pull/55454 finished to use in this function for the generic case.

From my perspective, that PR is good to go, just awaiting further comment or merging.

File: stdlib/Profile/src/Profile.jl [hunk 4]
```diff
@@ -990,7 +990,7 @@ function tree_format(frames::Vector{<:StackFrameTree}, level::Int, cols::Int, ma
                     link = editor_link(path, li.line)
                     styled"{link=$link:$styled_path}"
                 end
-                strs[i] = Base.annotatedstring(stroverhead, "╎", base, strcount, " ", rich_file, "; ", fname)
+                strs[i] = Base.annotatedstring(stroverhead, "╎", base, strcount, " ", rich_file, "  ", fname)
```
Comment (vtjnash) [inline stdlib/Profile/src/Profile.jl:993]: This commit broke BenchmarkTools. What is the intended fix, here or in BenchmarkTools?
https://github.com/JuliaCI/BenchmarkTools.jl/actions/runs/11898803449/job/33156070523

Comment (IanButterworth) [reply stdlib/Profile/src/Profile.jl:993]: Replace the `;` with a space in the regex in BenchmarkTools?

**Pull request merged**